File: packages/dart_node_react/test/ui.test.js
/**
 * UI Interaction Tests for Dart React Components
 * Uses React Testing Library for real DOM testing
 */

const { render, screen, fireEvent, waitFor } = require('@testing-library/react');
const userEvent = require('@testing-library/user-event').default;

// Make testing library available globally for Dart interop
global.testingLibrary = { render, screen, fireEvent, waitFor };
global.userEvent = userEvent;

// Load compiled Dart code
require('./dist/ui_test.js');

// Get exported Dart components
const {
  counterComponent,
  formComponent,
  toggleComponent,
  listComponent,
} = global.dartReactTests;

// =============================================================================
// Counter Component Tests
// =============================================================================

describe('Counter Component', () => {
  test('renders with initial count of 0', () => {
    render(counterComponent());

    const display = screen.getByTestId('count-display');
    expect(display).toHaveTextContent('Count: 0');
  });

  test('increments count when increment button clicked', async () => {
    const user = userEvent.setup();
    render(counterComponent());

    const incrementBtn = screen.getByTestId('increment-btn');
    const display = screen.getByTestId('count-display');

    await user.click(incrementBtn);
    expect(display).toHaveTextContent('Count: 1');

    await user.click(incrementBtn);
    expect(display).toHaveTextContent('Count: 2');

    await user.click(incrementBtn);
    expect(display).toHaveTextContent('Count: 3');
  });

  test('decrements count when decrement button clicked', async () => {
    const user = userEvent.setup();
    render(counterComponent());

    const decrementBtn = screen.getByTestId('decrement-btn');
    const display = screen.getByTestId('count-display');

    await user.click(decrementBtn);
    expect(display).toHaveTextContent('Count: -1');

    await user.click(decrementBtn);
    expect(display).toHaveTextContent('Count: -2');
  });

  test('resets count to 0 when reset button clicked', async () => {
    const user = userEvent.setup();
    render(counterComponent());

    const incrementBtn = screen.getByTestId('increment-btn');
    const resetBtn = screen.getByTestId('reset-btn');
    const display = screen.getByTestId('count-display');

    // Increment a few times
    await user.click(incrementBtn);
    await user.click(incrementBtn);
    await user.click(incrementBtn);
    expect(display).toHaveTextContent('Count: 3');

    // Reset
    await user.click(resetBtn);
    expect(display).toHaveTextContent('Count: 0');
  });

  test('handles rapid clicks correctly', async () => {
    const user = userEvent.setup();
    render(counterComponent());

    const incrementBtn = screen.getByTestId('increment-btn');
    const display = screen.getByTestId('count-display');

    // Rapid clicks
    for (let i = 0; i < 10; i++) {
      await user.click(incrementBtn);
    }

    expect(display).toHaveTextContent('Count: 10');
  });
});

// =============================================================================
// Form Component Tests
// =============================================================================

describe('Form Component', () => {
  test('renders empty form inputs', () => {
    render(formComponent(null));

    const emailInput = screen.getByTestId('email-input');
    const passwordInput = screen.getByTestId('password-input');

    expect(emailInput).toHaveValue('');
    expect(passwordInput).toHaveValue('');
  });

  test('updates email input when typing', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const emailInput = screen.getByTestId('email-input');

    await user.type(emailInput, 'test@example.com');
    expect(emailInput).toHaveValue('test@example.com');
  });

  test('updates password input when typing', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const passwordInput = screen.getByTestId('password-input');

    await user.type(passwordInput, 'secretpass123');
    expect(passwordInput).toHaveValue('secretpass123');
  });

  test('shows error when submitting empty form', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const submitBtn = screen.getByTestId('submit-btn');

    await user.click(submitBtn);

    const errorMessage = screen.getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows error when only email is filled', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const emailInput = screen.getByTestId('email-input');
    const submitBtn = screen.getByTestId('submit-btn');

    await user.type(emailInput, 'test@example.com');
    await user.click(submitBtn);

    const errorMessage = screen.getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows error when only password is filled', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const passwordInput = screen.getByTestId('password-input');
    const submitBtn = screen.getByTestId('submit-btn');

    await user.type(passwordInput, 'secret123');
    await user.click(submitBtn);

    const errorMessage = screen.getByTestId('error-message');
    expect(errorMessage).toHaveTextContent('All fields required');
  });

  test('shows success message when form is valid', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const emailInput = screen.getByTestId('email-input');
    const passwordInput = screen.getByTestId('password-input');
    const submitBtn = screen.getByTestId('submit-btn');

    await user.type(emailInput, 'test@example.com');
    await user.type(passwordInput, 'secret123');
    await user.click(submitBtn);

    const successMessage = screen.getByTestId('success-message');
    expect(successMessage).toHaveTextContent('Form submitted!');

    // Error should not be visible
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
  });

  test('calls onSubmit callback with form values', async () => {
    const user = userEvent.setup();
    const onSubmit = jest.fn();

    render(formComponent(onSubmit));

    const emailInput = screen.getByTestId('email-input');
    const passwordInput = screen.getByTestId('password-input');
    const submitBtn = screen.getByTestId('submit-btn');

    await user.type(emailInput, 'user@test.com');
    await user.type(passwordInput, 'mypassword');
    await user.click(submitBtn);

    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenCalledWith('user@test.com', 'mypassword');
  });

  test('clears error when valid form is submitted', async () => {
    const user = userEvent.setup();
    render(formComponent(null));

    const emailInput = screen.getByTestId('email-input');
    const passwordInput = screen.getByTestId('password-input');
    const submitBtn = screen.getByTestId('submit-btn');

    // First submit empty form to show error
    await user.click(submitBtn);
    expect(screen.getByTestId('error-message')).toBeInTheDocument();

    // Now fill and submit
    await user.type(emailInput, 'test@example.com');
    await user.type(passwordInput, 'password123');
    await user.click(submitBtn);

    // Error should be gone, success should show
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
    expect(screen.getByTestId('success-message')).toBeInTheDocument();
  });
});

// =============================================================================
// Toggle Component Tests
// =============================================================================

describe('Toggle Component', () => {
  test('renders with OFF state initially', () => {
    render(toggleComponent());

    const status = screen.getByTestId('toggle-status');
    expect(status).toHaveTextContent('OFF');
  });

  test('toggles to ON when button clicked', async () => {
    const user = userEvent.setup();
    render(toggleComponent());

    const toggleBtn = screen.getByTestId('toggle-btn');
    const status = screen.getByTestId('toggle-status');

    await user.click(toggleBtn);
    expect(status).toHaveTextContent('ON');
  });

  test('toggles back to OFF when clicked again', async () => {
    const user = userEvent.setup();
    render(toggleComponent());

    const toggleBtn = screen.getByTestId('toggle-btn');
    const status = screen.getByTestId('toggle-status');

    await user.click(toggleBtn);
    expect(status).toHaveTextContent('ON');

    await user.click(toggleBtn);
    expect(status).toHaveTextContent('OFF');
  });

  test('handles multiple toggles correctly', async () => {
    const user = userEvent.setup();
    render(toggleComponent());

    const toggleBtn = screen.getByTestId('toggle-btn');
    const status = screen.getByTestId('toggle-status');

    // Toggle 5 times
    await user.click(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');

    await user.click(toggleBtn); // OFF
    expect(status).toHaveTextContent('OFF');

    await user.click(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');

    await user.click(toggleBtn); // OFF
    expect(status).toHaveTextContent('OFF');

    await user.click(toggleBtn); // ON
    expect(status).toHaveTextContent('ON');
  });
});

// =============================================================================
// List Component Tests
// =============================================================================

describe('List Component', () => {
  test('renders empty list with message', () => {
    render(listComponent());

    const emptyMessage = screen.getByTestId('empty-message');
    expect(emptyMessage).toHaveTextContent('No items');

    const count = screen.getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 0');
  });

  test('adds item to list when add button clicked', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, 'First item');
    await user.click(addBtn);

    const item = screen.getByTestId('list-item-0');
    expect(item).toHaveTextContent('First item');

    const count = screen.getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 1');

    // Empty message should be gone
    expect(screen.queryByTestId('empty-message')).not.toBeInTheDocument();
  });

  test('clears input after adding item', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, 'Test item');
    await user.click(addBtn);

    expect(input).toHaveValue('');
  });

  test('does not add empty items', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const addBtn = screen.getByTestId('add-item-btn');

    await user.click(addBtn);

    // Should still show empty message
    expect(screen.getByTestId('empty-message')).toBeInTheDocument();

    const count = screen.getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 0');
  });

  test('does not add whitespace-only items', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, '   ');
    await user.click(addBtn);

    expect(screen.getByTestId('empty-message')).toBeInTheDocument();
  });

  test('adds multiple items correctly', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, 'Item 1');
    await user.click(addBtn);

    await user.type(input, 'Item 2');
    await user.click(addBtn);

    await user.type(input, 'Item 3');
    await user.click(addBtn);

    expect(screen.getByTestId('list-item-0')).toHaveTextContent('Item 1');
    expect(screen.getByTestId('list-item-1')).toHaveTextContent('Item 2');
    expect(screen.getByTestId('list-item-2')).toHaveTextContent('Item 3');

    const count = screen.getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 3');
  });

  test('removes item when remove button clicked', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, 'Item to remove');
    await user.click(addBtn);

    expect(screen.getByTestId('list-item-0')).toBeInTheDocument();

    const removeBtn = screen.getByTestId('remove-btn-0');
    await user.click(removeBtn);

    expect(screen.queryByTestId('list-item-0')).not.toBeInTheDocument();
    expect(screen.getByTestId('empty-message')).toBeInTheDocument();
  });

  test('removes correct item from middle of list', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    await user.type(input, 'First');
    await user.click(addBtn);

    await user.type(input, 'Second');
    await user.click(addBtn);

    await user.type(input, 'Third');
    await user.click(addBtn);

    // Remove middle item
    const removeBtn = screen.getByTestId('remove-btn-1');
    await user.click(removeBtn);

    // Check remaining items
    expect(screen.getByTestId('list-item-0')).toHaveTextContent('First');
    expect(screen.getByTestId('list-item-1')).toHaveTextContent('Third');
    expect(screen.queryByTestId('list-item-2')).not.toBeInTheDocument();

    const count = screen.getByTestId('items-count');
    expect(count).toHaveTextContent('Total: 2');
  });

  test('handles add and remove operations in sequence', async () => {
    const user = userEvent.setup();
    render(listComponent());

    const input = screen.getByTestId('add-item-input');
    const addBtn = screen.getByTestId('add-item-btn');

    // Add 3 items
    await user.type(input, 'A');
    await user.click(addBtn);
    await user.type(input, 'B');
    await user.click(addBtn);
    await user.type(input, 'C');
    await user.click(addBtn);

    expect(screen.getByTestId('items-count')).toHaveTextContent('Total: 3');

    // Remove first
    await user.click(screen.getByTestId('remove-btn-0'));
    expect(screen.getByTestId('items-count')).toHaveTextContent('Total: 2');

    // Add new item
    await user.type(input, 'D');
    await user.click(addBtn);
    expect(screen.getByTestId('items-count')).toHaveTextContent('Total: 3');

    // Remove all
    await user.click(screen.getByTestId('remove-btn-0'));
    await user.click(screen.getByTestId('remove-btn-0'));
    await user.click(screen.getByTestId('remove-btn-0'));

    expect(screen.getByTestId('empty-message')).toBeInTheDocument();
    expect(screen.getByTestId('items-count')).toHaveTextContent('Total: 0');
  });
});

File: packages/dart_node_react/test/ui_test.dart
/// UI interaction tests for React components
/// Compiled to JS and run with Jest + React Testing Library
///
/// Build: dart compile js test/ui_test.dart -o test/dist/ui_test.js
/// Run: cd test && npm test
library;

import 'dart:js_interop';
import 'dart:js_interop_unsafe';

import 'package:dart_node_react/dart_node_react.dart';

// =============================================================================
// Testing Library Bindings
// =============================================================================

@JS('testingLibrary.render')
external JSObject _render(JSObject element);

@JS('testingLibrary.screen')
external JSObject get _screen;

@JS('testingLibrary.fireEvent')
external JSObject get _fireEvent;

@JS('userEvent.setup')
external JSObject _userEventSetup();

// =============================================================================
// Test Utilities - Exported for shared use
// =============================================================================

/// Render a React element for testing
JSObject render(JSObject element) => _render(element);

/// Screen queries
JSObject get screen => _screen;

/// Fire events on elements
JSObject get fireEvent => _fireEvent;

/// User event instance for realistic interactions
JSObject userEvent() => _userEventSetup();

/// Query by text
JSObject? queryByText(String text) =>
    screen.callMethod('queryByText'.toJS, text.toJS) as JSObject?;

/// Get by text (throws if not found)
JSObject getByText(String text) =>
    screen.callMethod('getByText'.toJS, text.toJS)! as JSObject;

/// Query by placeholder
JSObject? queryByPlaceholder(String text) =>
    screen.callMethod('queryByPlaceholderText'.toJS, text.toJS) as JSObject?;

/// Get by placeholder (throws if not found)
JSObject getByPlaceholder(String text) =>
    screen.callMethod('getByPlaceholderText'.toJS, text.toJS)! as JSObject;

/// Query by role
JSObject? queryByRole(String role, [JSObject? options]) => (options != null)
    ? screen.callMethod('queryByRole'.toJS, role.toJS, options) as JSObject?
    : screen.callMethod('queryByRole'.toJS, role.toJS) as JSObject?;

/// Get by role (throws if not found)
JSObject getByRole(String role, [JSObject? options]) => (options != null)
    ? screen.callMethod('getByRole'.toJS, role.toJS, options)! as JSObject
    : screen.callMethod('getByRole'.toJS, role.toJS)! as JSObject;

/// Query by test ID
JSObject? queryByTestId(String testId) =>
    screen.callMethod('queryByTestId'.toJS, testId.toJS) as JSObject?;

/// Get by test ID (throws if not found)
JSObject getByTestId(String testId) =>
    switch (screen.callMethod('getByTestId'.toJS, testId.toJS)) {
      final JSObject obj => obj,
      _ => throw StateError('Element with testId "$testId" not found'),
    };

/// Click an element
void click(JSObject element) {
  fireEvent.callMethod('click'.toJS, element);
}

/// Change input value (sync)
void changeValue(JSObject element, String value) {
  final eventInit = JSObject()
    ..setProperty(
      'target'.toJS,
      JSObject()..setProperty('value'.toJS, value.toJS),
    );
  fireEvent.callMethod('change'.toJS, element, eventInit);
}

// =============================================================================
// Test Components
// =============================================================================

/// Simple counter component for testing state updates
JSObject counterComponent() => createElement(
  ((JSAny props) {
    final (countState, setCount) = useState(0.toJS);
    final count = (countState as JSNumber?)?.toDartInt ?? 0;

    return div(
      children: [
        span('Count: $count', props: {'data-testid': 'count-display'}),
        button(
          text: 'Increment',
          onClick: () => setCount.callAsFunction(null, (count + 1).toJS),
          props: {'data-testid': 'increment-btn'},
        ),
        button(
          text: 'Decrement',
          onClick: () => setCount.callAsFunction(null, (count - 1).toJS),
          props: {'data-testid': 'decrement-btn'},
        ),
        button(
          text: 'Reset',
          onClick: () => setCount.callAsFunction(null, 0.toJS),
          props: {'data-testid': 'reset-btn'},
        ),
      ],
    );
  }).toJS,
);

/// Form component for testing input interactions
JSObject formComponent({
  void Function(String email, String password)? onSubmit,
}) => createElement(
  ((JSAny props) {
    final (emailState, setEmail) = useState(''.toJS);
    final (passState, setPass) = useState(''.toJS);
    final (errorState, setError) = useState(null);
    final (submittedState, setSubmitted) = useState(false.toJS);

    final email = (emailState as JSString?)?.toDart ?? '';
    final password = (passState as JSString?)?.toDart ?? '';
    final error = (errorState as JSString?)?.toDart;
    final submitted = (submittedState as JSBoolean?)?.toDart ?? false;

    void handleSubmit() {
      (email.isEmpty || password.isEmpty)
          ? setError.callAsFunction(null, 'All fields required'.toJS)
          : () {
              setError.callAsFunction();
              setSubmitted.callAsFunction(null, true.toJS);
              onSubmit?.call(email, password);
            }();
    }

    return div(
      props: {'data-testid': 'form-container'},
      children: [
        if (error != null)
          div(props: {'data-testid': 'error-message'}, child: span(error))
        else
          span(''),
        if (submitted)
          div(
            props: {'data-testid': 'success-message'},
            child: span('Form submitted!'),
          )
        else
          span(''),
        div(
          children: [
            input(
              type: 'email',
              placeholder: 'Enter email',
              value: email,
              onChange: (e) =>
                  setEmail.callAsFunction(null, _extractInputValue(e)),
              props: {'data-testid': 'email-input'},
            ),
          ],
        ),
        div(
          children: [
            input(
              type: 'password',
              placeholder: 'Enter password',
              value: password,
              onChange: (e) =>
                  setPass.callAsFunction(null, _extractInputValue(e)),
              props: {'data-testid': 'password-input'},
            ),
          ],
        ),
        button(
          text: 'Submit',
          onClick: handleSubmit,
          props: {'data-testid': 'submit-btn'},
        ),
      ],
    );
  }).toJS,
);

JSString _extractInputValue(JSAny event) {
  final obj = event as JSObject;
  final target = obj['target'] as JSObject?;
  return (target?['value'] as JSString?) ?? ''.toJS;
}

/// Toggle component for testing boolean state
JSObject toggleComponent() => createElement(
  ((JSAny props) {
    final (onState, setOn) = useState(false.toJS);
    final isOn = (onState as JSBoolean?)?.toDart ?? false;

    return div(
      children: [
        span(isOn ? 'ON' : 'OFF', props: {'data-testid': 'toggle-status'}),
        button(
          text: 'Toggle',
          onClick: () => setOn.callAsFunction(null, (!isOn).toJS),
          props: {'data-testid': 'toggle-btn'},
        ),
      ],
    );
  }).toJS,
);

/// List component for testing dynamic rendering
JSObject listComponent() => createElement(
  ((JSAny props) {
    final (itemsState, setItems) = useState(<JSAny>[].toJS);
    final (inputState, setInput) = useState(''.toJS);

    final items = (itemsState as JSArray?)?.toDart ?? [];
    final inputValue = (inputState as JSString?)?.toDart ?? '';

    void addItem() => switch (inputValue.trim().isNotEmpty) {
      true => () {
        final newItems = [...items, inputValue.toJS];
        setItems.callAsFunction(null, newItems.toJS);
        setInput.callAsFunction(null, ''.toJS);
      }(),
      false => null,
    };

    void removeItem(int index) {
      final newItems = [...items]..removeAt(index);
      setItems.callAsFunction(null, newItems.toJS);
    }

    return div(
      props: {'data-testid': 'list-container'},
      children: [
        div(
          children: [
            input(
              type: 'text',
              placeholder: 'Add item',
              value: inputValue,
              onChange: (e) =>
                  setInput.callAsFunction(null, _extractInputValue(e)),
              props: {'data-testid': 'add-item-input'},
            ),
            button(
              text: 'Add',
              onClick: addItem,
              props: {'data-testid': 'add-item-btn'},
            ),
          ],
        ),
        div(
          props: {'data-testid': 'items-list'},
          children: items.isEmpty
              ? [
                  span('No items', props: {'data-testid': 'empty-message'}),
                ]
              : items
                    .asMap()
                    .entries
                    .map(
                      (entry) => div(
                        props: {
                          'key': entry.key,
                          'data-testid': 'list-item-${entry.key}',
                        },
                        children: [
                          span(switch (entry.value) {
                            final JSString s => s.toDart,
                            _ => '',
                          }),
                          button(
                            text: 'Remove',
                            onClick: () => removeItem(entry.key),
                            props: {'data-testid': 'remove-btn-${entry.key}'},
                          ),
                        ],
                      ),
                    )
                    .toList(),
        ),
        span('Total: ${items.length}', props: {'data-testid': 'items-count'}),
      ],
    );
  }).toJS,
);

// =============================================================================
// Export test components for Jest
// =============================================================================

@JS('dartReactTests')
external set _dartReactTests(JSObject value);

/// Wrapper to handle JS undefined vs null for formComponent
JSObject _wrapFormComponent(JSAny? onSubmit) => switch (onSubmit) {
      null => formComponent(),
      final cb when cb.isUndefinedOrNull => formComponent(),
      final JSFunction cb => formComponent(
          onSubmit: (email, password) =>
              cb.callAsFunction(null, email.toJS, password.toJS),
        ),
      _ => formComponent(),
    };

void main() {
  // Create formComponent wrapper that handles both 0 and 1 args
  final formComponentWrapper = ((JSAny? arg) =>
      (arg == null || arg.isUndefinedOrNull)
          ? formComponent()
          : _wrapFormComponent(arg)).toJS;

  _dartReactTests = JSObject()
    // Export components
    ..setProperty('counterComponent'.toJS, counterComponent.toJS)
    ..setProperty('formComponent'.toJS, formComponentWrapper)
    ..setProperty('toggleComponent'.toJS, toggleComponent.toJS)
    ..setProperty('listComponent'.toJS, listComponent.toJS)
    // Export utilities
    ..setProperty('render'.toJS, render.toJS)
    ..setProperty('getByText'.toJS, getByText.toJS)
    ..setProperty('getByTestId'.toJS, getByTestId.toJS)
    ..setProperty('getByPlaceholder'.toJS, getByPlaceholder.toJS)
    ..setProperty('queryByPlaceholder'.toJS, queryByPlaceholder.toJS)
    ..setProperty('queryByTestId'.toJS, queryByTestId.toJS)
    ..setProperty('queryByRole'.toJS, queryByRole.toJS)
    ..setProperty('getByRole'.toJS, getByRole.toJS)
    ..setProperty('click'.toJS, click.toJS)
    ..setProperty('changeValue'.toJS, changeValue.toJS)
    ..setProperty('queryByText'.toJS, queryByText.toJS)
    ..setProperty('userEvent'.toJS, userEvent.toJS);
}
